src/nvim/event/rstream.c

```diff
@@ -167,9 +169,19 @@ static void fread_idle_cb(uv_idle_t *handle)
   invoke_read_cb(stream, false);
 }
 
-static void invoke_read_cb(Stream *stream, bool eof)
+static void read_event(void **argv)
 {
+  Stream *stream = argv[0];
+  bool eof = (uintptr_t)argv[1];
   if (stream->read_cb) {
     stream->read_cb(stream, stream->buffer, stream->data, eof);
   }
+  if (eof && stream->internal_eof_cb) {
+    stream->internal_eof_cb(stream, stream->internal_data);
+  }
+}
+
+static void invoke_read_cb(Stream *stream, bool eof)
+{
+  queue_put(stream->events, read_event, 2, stream, (void *)(uintptr_t)eof);
```

Comment:
`on_job_output()` in eval.c consumes all data in the `RBbuffer` (`RBUFFER_UNTIL_EMPTY`). So if we queue more than one `read_event` for a stream, the first event will use up all data. The other events only see the empty `RBbuffer`. This could lead to reordering of output for different streams if their read events are interleaved.